### PR TITLE
Remove `balance` and `devices` parameter from Pipe.

### PR DESCRIPTION
**Pull request by pritamdamania87**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #47829 Modify Pipe to return an RRef.
* **#46804 Remove `balance` and `devices` parameter from Pipe.**

As per our design in https://github.com/pytorch/pytorch/issues/44827,
changign the API such that the user places modules on appropriate devices
instead of having a `balance` and `devices` parameter that decides this.

This design allows us to use RemoteModule in the future.

Differential Revision: [D24524219](https://our.internmc.facebook.com/intern/diff/D24524219/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24524219/)!

### CODE REVIEW

File: test/distributed/_pipeline/sync/test_pipe.py [hunk 1]
```diff
@@ -525,46 +485,55 @@ def test_named_children():
...
+    model = nn.Sequential(Surrogate(conv1, conv2.cuda()))
+
+    with pytest.raises(
+        ValueError,
+        match='should have all parameters on a single device, please use .to\(\)'
+            ' to place the module on a single device'):
+        Pipe(model)
+
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="cuda required")
+@pytest.mark.skipif(torch.cuda.device_count() < 2, reason="Need atleast two GPUs")
```
Comment (mrshenli) [inline test/distributed/_pipeline/sync/test_pipe.py:520]: curious, doesn't existing unittest skip helpers work? I did use those locally with pytest and seems working for me.

File: torch/distributed/_pipeline/sync/pipe.py [hunk 2]
```diff
@@ -89,66 +89,46 @@ class BalanceError(ValueError):
...
-) -> Tuple[List[nn.Sequential], List[int], List[torch.device]]:
-    """Splits a module into multiple partitions.
+def _retrieve_device(module: nn.Module) -> torch.device:
+    """Validates all parameters in the Module have the same device and returns
+    the appropriate device.
 
-    Returns:
-        A tuple of (partitions, balance, devices).
+    Arguments:
+        An nn.Module to process.
```
Comment (mrshenli) [inline torch/distributed/_pipeline/sync/pipe.py:97]: double quote code

File: torch/distributed/_pipeline/sync/pipe.py [hunk 3]
```diff
@@ -89,66 +89,46 @@ class BalanceError(ValueError):
...
 
-        Partitions are represented as a :class:`~torch.nn.ModuleList` whose
-        item is a partition. All layers in a partition are placed in the
-        same device.
+    Returns:
+        torch.Device for the entire module.
```
Comment (mrshenli) [inline torch/distributed/_pipeline/sync/pipe.py:100]: ditto

File: torch/distributed/_pipeline/sync/pipe.py [hunk 4]
```diff
@@ -89,66 +89,46 @@ class BalanceError(ValueError):
...
 
     Raises:
-        BalanceError:
-            wrong balance
-        IndexError:
-            the number of devices is fewer than the number of partitions.
-
+        ValueError:
+            If devices for nn.Module parameters are not all same.
```
Comment (mrshenli) [inline torch/distributed/_pipeline/sync/pipe.py:104]: ditto

File: torch/distributed/_pipeline/sync/pipe.py [hunk 5]
```diff
@@ -160,28 +140,27 @@ class Pipe(Module):
...
 
     Pipe combines pipeline parallelism with checkpointing to reduce peak
     memory required to train while minimizing device under-utilization.
 
-    You should determine the balance when defining a :class:`Pipe` module, as
-    balancing will not be done automatically. The module will be partitioned
-    into multiple devices according to the given balance. You may rely on
-    heuristics to find your own optimal configuration.
+    You should place all the modules on the appropriate devices before passing
+    them to this API and wrap them into an nn.Sequential module defining the
```
Comment (mrshenli) [inline torch/distributed/_pipeline/sync/pipe.py:152]: ditto

File: torch/distributed/_pipeline/sync/pipe.py [hunk 6]
```diff
@@ -160,28 +140,27 @@ class Pipe(Module):
...
 
     Args:
         module (torch.nn.Sequential):
-            sequential module to be parallelized
-        balance (ints):
-            list of number of layers in each partition
+            Sequential module to be parallelized using pipelining. Each module
+            in the sequence has to have all of its parameters on a single
+            device. Each module in the sequence has to either be an nn.Module
+            or nn.Sequential (to combine multiple sequential modules on a single
```
Comment (mrshenli) [inline torch/distributed/_pipeline/sync/pipe.py:160]: ditto for this block

File: torch/testing/_internal/distributed/pipeline/utils.py [hunk 7]
```diff
@@ -0,0 +1,21 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All rights reserved.
+#
+# This source code is licensed under the BSD license found in the
+# LICENSE file in the root directory of this source tree.
+
+from torch import nn
+from typing import List
+
+def convert_to_balance(pipe: nn.Sequential, balance: List[int]):
```
Comment (mrshenli) [inline torch/testing/_internal/distributed/pipeline/utils.py:9]: any reason for having a second `convert_to_balance` in test folder?

**Review by mrshenli — APPROVED** [review]
LGTM! Added some minor comments.

**Comment by ngimel** [comment]
Reverting, broke flake8

**Pull request closed without merging**